articles/flask-crud-api/index.md

```diff
@@ -0,0 +1,360 @@
+
+### Project configuration
+
+In this section, we create an instance of the flask application, setup the database configurations and create a database object.
+
+```python
+
+app = Flask(__name__)
+
+app.config['SQLALCHEMY_DATABASE_URI'] = 'mysql+pymysql://root:@localhost:3306/todo'
+
+db = SQLAlchemy(app)
+
+```
+
+- `app = Flask(__name__)` creates an instance of the flask web application.
+- `app.config['SQLALCHEMY_DATABASE_URI'] = 'mysql+pymysql://<username>:<password>@localhost:
+<database port>/<database name>'
+` contains the information required to connect to our database. Our database username is `root` ,the password is an empty string ` ` ,the database port is `3306` and the database name is `todo`.
+
+On your computer start the XAMPP control panel, start the apache and MySQL servers.
+On your browser open `http://localhost/dashboard/` then `phpAdmin`, create a database with the name `todo` as shown below.
+
+![Creating database](/engineering-education/flask-crud-api/database.png)
+
+`db = SQLAlchemy(app)` creates an object of the SQLAlchemy and stores it in a variable `db`.
+
+### Creating the Todo model
+
+We are going to be creating the `Todo` model which will be a Python class representing the `Todo` table in the database.
+
+```python
+
+# Model
+
+class Todo(db.Model):
+
+   __tablename__ = "todos"
+
+   id = db.Column(db.Integer, primary_key=True)
+
+   title = db.Column(db.String(20))
+
+   todo_description = db.Column(db.String(100))
+
+   def create(self):
+
+       db.session.add(self)
+
+       db.session.commit()
+
+       return self
+
+   def __init__(self, title, todo_description):
+
+       self.title = title
+
+       self.todo_description = todo_description
+
+   def __repr__(self):
+
+       return f"{self.id}"
+
+db.create_all()
+
+```
+
+`Todo` model class has an id property which is used as the primary key.
+
+- `db.create_all()` makes the application to create all the defined tables in the database.
+
+- `db.session.add(self)` adds the `Todo` instance into the SQLAlchemy database connection session.
+
+- `db.session.commit()` executes all the database operations that are available in the session.
+
+```python
+
+class TodoSchema(ModelSchema):
+
+   class Meta(ModelSchema.Meta):
+
+       model = Todo
+
+       sqla_session = db.session
+
+   id = fields.Number(dump_only=True)
+
+   title = fields.String(required=True)
+
+   todo_description = fields.String(required=True)
+
+```
+
+The `TodoSchema` class above makes it possible to return JSON from the Python objects from the SQLAlchemy.
+
+### Creating the API endpoints
+
+**POST endpoint**
+
+```python
+
+@app.route('/api/v1/todo', methods=['POST'])
+
+def create_todo():
+
+   data = request.get_json()
+
+   todo_schema = TodoSchema()
+
+   todo = todo_schema.load(data)
+
+   result = todo_schema.dump(todo.create())
+
+   return make_response(jsonify({"todo": result}), 200)
+
```

Comment:
the line `todo = todo_schema.load(data)` throws an error `marshmallow.exceptions.ValidationError: {'_schema': ['Invalid input type.']}` on a POST request at /api/v1/todo. Please update your code to fix this.

Reply:
@paulodhiambo, Everything is working fine. I got where the error was coming from. I was testing the API using form data... but it works fine with JSON. Sorry about that.

Reply:
@geoffrey45 seems you are passing in a list as your POST request body as shown below.
```JSON
[
    {
        "title": "Go to School",
        "todo_description": "Complete assignments at School"
    }
]
```
The POST endpoint only allows for a single `Todo` item at a time as shown below.
```JSON
{
    "title": "Go to School",
    "todo_description": "Complete assignments at School"
}
```
When a list is passed to the POST endpoint the below error occurs.
```bash
[2020-12-30 12:31:56,563] ERROR in app: Exception on /api/v1/todo [POST]
Traceback (most recent call last):
  File "/home/paulodhiambo/Public/tuts/trial/venv/lib/python3.8/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/paulodhiambo/Public/tuts/trial/venv/lib/python3.8/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/paulodhiambo/Public/tuts/trial/venv/lib/python3.8/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/paulodhiambo/Public/tuts/trial/venv/lib/python3.8/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/home/paulodhiambo/Public/tuts/trial/venv/lib/python3.8/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/paulodhiambo/Public/tuts/trial/venv/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/paulodhiambo/Public/tuts/trial/app.py", line 46, in create_todo
    todo = todo_schema.load(data)
  File "/home/paulodhiambo/Public/tuts/trial/venv/lib/python3.8/site-packages/marshmallow_sqlalchemy/schema/load_instance_mixin.py", line 92, in load
    return super().load(data, **kwargs)
  File "/home/paulodhiambo/Public/tuts/trial/venv/lib/python3.8/site-packages/marshmallow/schema.py", line 727, in load
    return self._do_load(
  File "/home/paulodhiambo/Public/tuts/trial/venv/lib/python3.8/site-packages/marshmallow/schema.py", line 909, in _do_load
    raise exc
marshmallow.exceptions.ValidationError: {'_schema': ['Invalid input type.']}
```